river/main.zig

```diff
@@ -126,39 +127,36 @@ pub fn main() anyerror!void {
         .warn, .err, .crit, .alert, .emerg => .err,
     });
 
-    log_server.info("initializing", .{});
-
     if (startup_command == null) {
-        if (try getStartupCommand()) |path| {
-            startup_command = path;
-            log_server.info("Using default startup command path: {}", .{path});
-        } else {
-            log_server.info("Starting without startup command", .{});
-        }
-    } else {
-        log_server.info("Using custom startup command path: {}", .{startup_command});
+        if (try getStartupCommand()) |path| startup_command = path;
     }
 
+    log_server.info("initializing server", .{});
     var server: Server = undefined;
     try server.init();
     defer server.deinit();
 
     try server.start();
 
-    const child_pid = if (startup_command) |cmd| blk: {
+    // Run the child in a new process group so that we can send SIGTERM to all
+    // descendants on exit.
+    const child_pgid = if (startup_command) |cmd| blk: {
+        log_server.info("running startup command '{}'", .{cmd});
         const child_args = [_:null]?[*:0]const u8{ "/bin/sh", "-c", cmd, null };
-        const pid = try std.os.fork();
+        const pid = try os.fork();
         if (pid == 0) {
-            if (std.os.system.sigprocmask(std.os.SIG_SETMASK, &std.os.empty_sigset, null) < 0) unreachable;
-            std.os.execveZ("/bin/sh", &child_args, std.c.environ) catch c._exit(1);
+            if (c.setsid() < 0) unreachable;
+            if (os.system.sigprocmask(os.SIG_SETMASK, &os.empty_sigset, null) < 0) unreachable;
+            os.execveZ("/bin/sh", &child_args, std.c.environ) catch c._exit(1);
```

Comment:
Since, this file is also touched. Maybe it makes sense to look up `sh` in user's `$PATH` as was discussed on IRC?

Reply:
I'd rather do that change in a separate PR, this one is large enough. Also I have yet to learn of a system that both can build river and does not have `/bin/sh`.

Reply:
NixOS

Reply:
> I'd rather do that change in a separate PR, this one is large enough.

Makes sense, but maybe then this patch don't need to touch this section? It doesn't seem to be related to layouts in any way.

> Also I have yet to learn of a system that both can build river and does not have `/bin/sh`.

Pretty sure that's NixOS

Reply:
> Makes sense, but maybe then this patch don't need to touch this section? It doesn't seem to be related to layouts in any way.

This is related, because I needed these changes to make using a service manager (runit) as my startup process ergonomic. This service manager is responsible for running `rivertile` among other things as `rivertile` now needs to run continuously as a separte process.

> Pretty sure that's NixOS

Pretty sure NixOS has `/bin/sh`, otherwise sway's `exec` commands wouldn't work.

Reply:
@yyp Also note that this PR contains in fact multiple commits ("patches"). While the history is not final yet (some commits probably should be squashed), this PR will definitely have multiple commits. This PR containing other changes that are required by or depend on the layout implementation but are not directly related is fine. As long as commits themselves are strictly on-topic, the git history will be clean regardless.

Also getting `sh` from `$PATH` wasn't *really* discussed on IRC AFAIK. It was just mind dumb when I was wondering about #263. I have since come to the conclusion that this would not fix the issue, so I no longer consider doing that a good idea. Despite `/bin/sh` apparently not being a defined standard, it exists on effectively any machine. Now you could argue that `/bin/sh` isn't guaranteed to be a POSIX compatible shell and I'd agree, but then it should explicitly configurable, not implicitly snatched from the users `$PATH`. And that really belongs into another PR, to keep the discussions on-topic.

Reply:
> > Makes sense, but maybe then this patch don't need to touch this section? It doesn't seem to be related to layouts in any way.
> 
> This is related, because I needed these changes to make using a service manager (runit) as my startup process ergonomic. This service manager is responsible for running `rivertile` among other things as `rivertile` now needs to run continuously as a separte process.

Fair enough

> > Pretty sure that's NixOS
> 
> Pretty sure NixOS has `/bin/sh`, otherwise sway's `exec` commands wouldn't work.

Maybe. But I still think that hard-coding `/bin/sh` is a bit of a weird move.